codecov_auth/views/okta.py

```diff
@@ -0,0 +1,191 @@
+
+        current_user = self._login_user(request, iss, user_data)
+
+        # TEMPORARY: we're assuming a single owner for the time being since there's
+        # no supporting UI to select which owner you'd like to view
+        owner = current_user.owners.first()
+        if owner is not None:
+            service = get_short_service_name(owner.service)
+            response = redirect(f"{settings.CODECOV_DASHBOARD_URL}/{service}")
+        else:
+            # user has not connected any owners yet
+            response = redirect(f"{settings.CODECOV_DASHBOARD_URL}/sync")
+
+        response.delete_cookie("_okta_iss")
+        return response
+
+    def _login_user(self, request: HttpRequest, iss: str, user_data: dict):
+        id_token = user_data[
+            "id_token"
+        ]  # this will be present since we requested the `oidc` scope
+        id_payload = validate_id_token(iss, id_token)
+
+        okta_id = id_payload["sub"]
+        user_email = id_payload["email"]
+        user_name = id_payload["name"]
+
+        okta_user = OktaUser.objects.filter(okta_id=okta_id).first()
+
+        current_user = None
+        if request.user is not None and not request.user.is_anonymous:
+            # we're already authenticated
+            current_user = request.user
+
+            if okta_user and okta_user.user != request.user:
+                log.warning(
+                    "Okta account already linked to another user",
+                    extra=dict(
+                        current_user_id=request.user.pk, okta_user_id=okta_user.pk
+                    ),
+                )
+                # Logout the current user and login the user who already
+                # claimed this Okta account (below)
+                logout(request)
+                current_user = okta_user.user
+        else:
+            # we're not authenticated
+            if okta_user:
+                log.info(
+                    "Existing Okta user logging in",
+                    extra=dict(okta_user_id=okta_user.pk),
+                )
+                current_user = okta_user.user
+            else:
+                current_user = User.objects.create(
+                    name=user_name,
+                    email=user_email,
+                )
+
+        if okta_user is None:
+            okta_user = OktaUser.objects.create(
+                user=current_user,
+                okta_id=okta_id,
+                name=user_name,
+                email=user_email,
+                access_token=user_data["access_token"],
+            )
+            log.info(
+                "Created Okta user",
+                extra=dict(okta_user_id=okta_user.pk),
+            )
+
+        login(request, current_user)
+        return current_user
+
+    def get(self, request):
+        if request.GET.get("code"):
+            return self._perform_login(request)
+        else:
+            iss = settings.OKTA_ISS or request.GET.get("iss")
```

Comment:
Optional: not sure if `iss` over `issuer` is understandable if you're familiar with OIDC, but otherwise I'd push for the latter for readability

Reply:
The query string param has to be `iss` since that's what Okta sends.  I think since it's referred to as `iss` elsewhere (i.e. in the ID token, query param, etc.) then I'm inclined to keep it as `iss` so you know it's the same thing.  I expect you'd need some familiarity with OIDC to fully follow this code anyway.

Reply:
Gotchaa, yeap that makes sense! We can keep it as is